### DISCUSSION TITLE
What certificate does libcurl use for windows compilation?

### DISCUSSION THREAD

**Discussion by sayoojbk** [curl/curl#8581] [Q&A] [answered]
I have to do a ssl connection to a server and i have a p12 file. I did the openssl approach to create cleint certificate and key and was able to connect and send my post request in my Linux compilation. But not able to use the same file for windows compilation. I also tried passing the certificate and key after importing using the certificate store into a folder project/libcurlssh in windows but still face error schannel failed to import cert file: 658383....xxx .

**Comment by jay** [comment, marked as answer]
Are you using [CURLOPT_SSLCERT](https://curl.se/libcurl/c/CURLOPT_SSLCERT.html) and have you reviewed the documentation? Schannel backend only accepts P12 certificates, or certificates in the OS certificate store if you use the thumbprint of the certificate.

In either case of Schannel, [CURLOPT_SSLKEY](https://curl.se/libcurl/c/CURLOPT_SSLKEY.html) (which is used to read a separate private key file) is ignored because the private key is expected to be in the client certificate specified in the certificate store or p12 file. I will update the documentation to mention that.

**Reply by sayoojbk** [reply]
Oh okay but does that mean the file has to be in the certificate store path not in a custom path for my project that i have created. Also didn't understand what you meant by sslkey is ignored in Linux we have two field right one is cert file and the other we pass to --key parameters and pardon me I'f I'm wrong that means one is client cert and the other client cert key

**Reply by sayoojbk** [reply]
And yes I'm using CURLOPT_SSLKEY.

**Reply by jay** [reply]
> Also didn't understand what you meant by sslkey is ignored in Linux

Sorry I was unclear. I meant in either case of Schannel, P12 certificates or certificates in the OS certificate store. I have [submitted a PR](https://github.com/curl/curl/pull/8587) that I hope explains it better.

**Reply by sayoojbk** [reply]
Hey @jay. That cleared my doubt. Also in case, it is possible can you add a sample code snippet for the windows use case with a certificate just like you have added for the Linux version. That will be really helpful in the future when someone stumbles upon this.

**Reply by jay** [reply]
I have explained it in the proposed documentation that the key file option is ignored by schannel since it uses P12 files which do not have a separate key file. I think that is sufficient.

**Comment by bagder** [comment]
> ssh connection

Did you really mean SSH? SSH doesn't typically use certificates at all.

**Reply by sayoojbk** [reply]
Sorry my bad ssl. That is an https call . Sorry don't have much knowledge about https, SSL and so but need to work on that for my task in using your curl library for our C++ Project ...

**Comment by Cmo1976** [comment]
I unblocked you sorry I have trust issues.I’m on an iOS system

On Thu, Mar 10, 2022 at 22:33 Sayooj ***@***.***> wrote:

> I have to do a ssh connection to a server and i have a p12 file. I did the
> openssl approach to create cleint certificate and key and was able to
> connect and send my post request in my Linux compilation. But not able to
> use the same file for windows compilation. I also tried passing the
> certificate and key after importing using the certificate store into a
> folder project/libcurlssh in windows but still face error schannel failed
> to import cert file: 658383....xxx .
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/curl/curl/discussions/8581>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXVTGU6JTKLVXWWFIPON6NTU7LSKBANCNFSM5QOXJIFQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Thank you,
Christine Heine

From: CMiHeine Affiliate
Christine Heine
